File: 2023/Leetcode/problem155.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.mins = []
        

    def push(self, val: int) -> None:
        if len(self.mins) == 0 or self.mins[-1] >= val:
            self.mins.append(val)
        self.stack.append(val)


    def pop(self) -> None:
        val = self.stack.pop()
        if val == self.mins[-1]:
            self.mins.pop()


    def top(self) -> int:
        return self.stack[-1]


    def getMin(self) -> int:
        return self.mins[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 2024/Leetcode/problem155.cpp
class MinStack {
private:
    vector<int> vals;
    vector<int> mins;
public:
    MinStack() {

    }
    
    void push(int val) {
        vals.push_back(val);
        if (mins.size() > 0) {
            mins.push_back(val > mins.back() ? mins.back() : val);
        } else {
            mins.push_back(val);
        }
    }
    
    void pop() {
        vals.pop_back();
        mins.pop_back();
    }
    
    int top() {
        return vals.back();
    }
    
    int getMin() {
        return mins.back();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */